lib/opencensus/trace/integrations/sidekiq_middleware.rb

```diff
@@ -0,0 +1,126 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require "opencensus"
+require "opencensus/trace/integrations/sidekiq"
+
+module OpenCensus
+  module Trace
+    module Integrations
+      # # Sidekiq integration
+      #
+      # This is a middleware for Sidekiq applications:
+      #
+      # * It wraps all jobs in a root span
+      # * It exports the captured spans at the end of the job
+      #
+      # ## Configuration
+      #
+      # Example:
+      # # config/initializers/sidekiq.rb
+      #
+      # require "opencensus/trace/integrations/sidekiq_middleware"
+      # Sidekiq.configure_server do |config|
+      #   config.server_middleware do |chain|
+      #     chain.add OpenCensus::Trace::Integrations::SidekiqMiddleware
+      #   end
+      # end
+      #
+      class SidekiqMiddleware
+        HTTP_HOST_ATTRIBUTE = "http.host".freeze
+        SEPARATOR = "/".freeze
+
+        ##
+        # Create the Sidekiq middleware.
+        #
+        # @param [#export] exporter The exported used to export captured spans
+        #     at the end of the request. Optional: If omitted, uses the exporter
+        #     in the current config.
+        #
+        def initialize exporter: nil
+          @exporter = exporter || OpenCensus::Trace.config.exporter
+
+          config = configuration
+          @trace_prefix = config.trace_prefix
+          @job_name_attrs = config.job_attrs_for_trace_name
+          @job_span_attrs = config.job_attrs_for_span
+        end
+
+        # @param [Object] worker the worker instance
+        # @param [Hash] job the full job payload
+        #   * @see https://github.com/mperham/sidekiq/wiki/Job-Format
+        # @param [String] queue the name of the queue the job was pulled from
+        # @yield the next middleware in the chain or worker `perform` method
+        # @return [Void]
+        def call _worker, job, _queue
+          trace_path = [@trace_prefix, job.values_at(*@job_name_attrs)]
+                       .join(SEPARATOR)
+
+          # TODO: find a way to give the job data to the sampler
+          # Duplicate this class maybe
+          #   lib/opencensus/trace/formatters/trace_context.rb
+          # trace_context: job.slice(*%w(class args queue)),
+
+          # TODO: use a sampler. We need to figure out how to pass job details
+          # to the sampler to choose whether or not to sample this run
+          unless configuration.sample_proc.call(job)
+            yield
+            return
+          end
+
+          Trace.start_request_trace do |span_context|
+            begin
+              Trace.in_span trace_path do |span|
+                start_job span, job.slice(*@job_span_attrs)
+                yield
```

Comment:
In my internal implementation, I would even capture the error of the job, then put some attributes for the span for easier filtering process. That's not crucial, just a suggestion ;)

Reply:
Interesting idea. I'll see if we can get some clarification on what process will be followed on features like this before the merger into the OpenTelemetry project.

@dazuma Can you tell us how you would like to handle open PRs in OpenCensus during the merge of the project into OpenTelemetry?

Reply:
Hi @dazuma, this PR has been open for a while. We're have implemented this branch on two of our internal apps in production and two more on the way. Are you happy to merge this PR, or do you have some feedback?